Build: Allow for file based deploy, sign packages

spinscale opened the pull request:
This allows for a local file based deploy without needed nexus
auth information.

Also signing of packages has been added, either via gradle.properties
or using system properties as a fallback.

The property build.repository allows to configure another endpoint if no
snapshot build is done.

rjernst commented on build.gradle:
```diff
@@ -50,21 +50,37 @@ subprojects {
+    nexus {
+      def gitProvider = new nebula.plugin.info.scm.GitScmProvider()
+      def shortHash = gitProvider.calculateChange(new File('.'))
+      def isRelease = System.getProperty('build.snapshot', 'true') == 'false'
```
Please use actual types, we've tried to avoid using def in the rest of gradle/groovy files here.

spinscale replied:
fixed

rjernst commented on build.gradle:
```diff
@@ -50,21 +50,37 @@ subprojects {
     }
+    nexus {
+      def gitProvider = new nebula.plugin.info.scm.GitScmProvider()
+      def shortHash = gitProvider.calculateChange(new File('.'))
```
The "scm provider" is something internal to the nebula info plugin. If you want the git hash, use jgit, which is what GitScmProvider uses.

spinscale replied:
changed, but what is the gain of this? If we exlcude the nebula plugin we also lose the jgit dependency, right? You just dont want to have code depending on this in the `build.gradle` file?

rjernst commented on build.gradle:
```diff
@@ -50,21 +50,37 @@ subprojects {
+    }
     // we have our own username/password prompts so that they only happen once
-    // TODO: add gpg signing prompts
+    // TODO: add gpg signing prompts, which is tricky, as the buildDeb/buildRpm tasks are executed before this code block
```
hrm, what do you mean by "executed"? This is all the configuration phase.

spinscale replied:
sorry for the confusion.. what I meant is that the configuration for the packages in `ospackage` is executed before this `whenReady` code block and thus reading a key here and applying it is tricky. I guess we can do some in-place replace magic?

rjernst replied:
Yes we can probably do some magic...it looks like the signing support was added ospackage in a non-standard way. Throughout gradle APIs, you see settings like this as Object so that they can be lazily evaluated. Here that would have allowed setting a placeholder object, which the impl of could have had the value filled in for.  Perhaps we should think about a PR back to ospackage to do that (change the types for the [signing key id and passphrase](https://github.com/nebula-plugins/gradle-ospackage-plugin/commit/0d484c764ce72dcb33f695e5de4f1f9ccb663bc6#diff-4c98103916724963b2b69c0918249f1eR36) to Object).

rjernst commented on distribution/rpm/build.gradle:
```diff
@@ -18,6 +18,13 @@
  */
 
 task buildRpm(type: Rpm) {
+  signingKeyId = project.hasProperty('signing.keyId') ? project.getProperty('signing.keyId') : System.getProperty('signing.keyId', 'D88E42B4')
```
Can't all of this be set in the ospackage section of distribution/build.gradle, so it applies to both deb and rpm?

spinscale replied:
good one, tried it, worked!

rjernst commented on distribution/build.gradle:
```diff
@@ -285,6 +285,14 @@ configure(subprojects.findAll { ['deb', 'rpm'].contains(it.name) }) {
     version project.version.replace('-SNAPSHOT', '')
 
+    // signing setup
+    signingKeyId = project.hasProperty('signing.keyId') ? project.getProperty('signing.keyId') : System.getProperty('signing.keyId', 'D88E42B4')
```
getProperty() is not a method on Project, it is just property(). Also, I don't think we should mess with both gradle properties and java system properties. We can just use gradle properties, passing them in like `-Psigning.password=foo`. Finally, I don't think we should set this default for signing.keyId. But we should have a default for the keyring file? See how the [nexus plugin](https://github.com/bmuschko/gradle-nexus-plugin/blob/master/src/main/groovy/com/bmuschko/gradle/nexus/NexusPlugin.groovy#L110) does it.

spinscale replied:
just using gradle properties now (I think we should port `build.snapshot` over as a gradle property as well, but not in this PR). Can you explain what is bad about setting our signing key as the default?

rjernst commented:
I looked into the issue about the asc file for tar.gz being named incorrectly. In order to fix it, in `distribution/tar/build.gradle`, remove the `archives buildTar` line, and add the following:

```
project.afterEvaluate {
  // gradle is completely broken for extensions that contain a dot, so we must be explicit
  project.signArchives.singleSignature.type = 'tar.gz.asc'
}
```

spinscale commented:
added the tar.gz fix as well, works. thx!

rjernst commented:
LGTM

spinscale committed:
Build: Allow for file based deploy, sign packages

This allows for a local file based deploy without needed nexus
auth information.

Also signing of packages has been added, either via gradle.properties
or using system properties as a fallback.

The property build.repository allows to configure another endpoint if no
snapshot build is done.

Fix creation of .asc file for tar.gz distribution

Closes #17405
